### DISCUSSION TITLE
命名的箭头函数（Named Arrow Function）

### DISCUSSION THREAD

**Discussion by JasinYip** [JSCIG/es-discuss#14] [Ideas]
为了程序的可读性，我们可能需要为箭头函数进行命名。

现在我们的做法与原来的 `function` 匿名函数一样，将箭头函数赋给一个标识符或对象键，如：

```js
const sum = (a, b) => a + b

const obj = {
  sum: (a, b) => a + b
}
```

但当我们在一个链式调用或 pipeline 中，我们就没法为函数命名了，如 Lodash Flow：

```js
const { flow } = require('lodash')

const result = flow([
  (a, b) => a + b,
])(1, 2)
```

在讨论中的 Pipeline（在此不展开讨论最后是否用这个方案）我们也有类似的问题：

```js
const result = "Hello"
  |> (str) => str + str
```

我们是否可能直接在箭头函数前面为此箭头函数命名？就像这样：

```js
// loadsh flow
const { flow } = require('lodash')

const result = flow([
  sum(a, b) => a + b
])

// pipeline
const result = "Hello"
  |> duplicate(str) => str + str
```

这个是否有较高的技术实现上的难度或设计上的冲突？

**Comment by Foolyou** [comment]
有 `async () => {}` 存在了，这个语法会破坏一致性，不如支持下：
`function (^name, ...params) {}`
和 `(^name, ...params) => {}` ？
（逃

**Comment by JasinYip** [comment]
> 有 async () => {} 存在了，这个语法会破坏一致性

这样是否可以？

```js
async sum (a, b) => { /* do some async works */ }
```

**Comment by Foolyou** [comment]
> > 有 async () => {} 存在了，这个语法会破坏一致性
> 
> 这样是否可以？
> 
> ```js
> async sum () => {} 
> ```

我的意思是，你还需要区分 `async () => {}` 和 `h () => {}`，当然解析的时候可以把 `async` 当成特例，但是这似乎对 parser 和阅读者都不是很友好。

**Comment by JasinYip** [comment]
感觉这个还好吧，`async` 可以说是一种事实上函数修饰符了，像关键词一样了。

**Comment by Huxpro** [comment]
其实 `async` 无所谓就是个 LL(1)，但是 `foo() => {...}` 和普通函数调用 `foo()` 是个 LL(k)，不过这个可能也不算什么大问题，毕竟 `() => {}` 跟 `(...)` 也是 LL(k) ……

所以具体为什么不支持可能还需要考证（

**Comment by Alan-Liang** [comment]
其实我希望有 named lambda 的主要原因是方便自引用。

```javascript
const callback = () => {
  console.log(Date.now() - this.startTime)
  setTimeout(callback, 1000)
}
setTimeout(callback, 10000)
```

v.s.

```javascript
const ctx = this
setTimeout(function callback () {
  console.log(Date.now() - ctx.startTime)
  setTimeout(callback, 1000)
}, 10000)
```

v.s.

```javascript
setTimeout(callback () => {
  console.log(Date.now() - this.startTime)
  setTimeout(callback, 1000)
}, 10000)
```

**Reply by magic-akari** [reply]
v.s. `Z combinator`

```javascript
const Z = (f) => ((x) => f((y) => x(x)(y)))((x) => f((y) => x(x)(y)));

const startTime = Date.now();

setTimeout(
  Z((callback) => () => {
    console.log(Date.now() - startTime);
    setTimeout(callback, 1000);
  }),
  10000
);
```

**Comment by JasinYip** [comment]
@Alan-Liang 没错，自引用或者叫递归调用，也是一个能被优化的场景

**Comment by Alan-Liang** [comment]
递归的话在尾递归优化落实之前还是算了吧😂 随便写点什么都能爆栈……

**Comment by JasinYip** [comment]
> 递归的话在尾递归优化落实之前还是算了吧😂 随便写点什么都能爆栈……

那倒不至于，很多场景还是需要递归的，比如遍历树、图等，爆栈属于自己代码写得有问题吧。

这可能不属于这个话题下讨论的东西。

**Comment by cloudyan** [comment]
期望支持，错误便于跟踪

**Comment by hwangweihang** [comment]
谢谢，信已收到。

**Comment by hwangweihang** [comment]
谢谢，信已收到。

**Comment by hax** [comment]
杨健找到了一些老的讨论：https://esdiscuss.org/topic/named-arrow-functions

看起来主要问题还是当初命名箭头函数在语法上和 `async () => {}` 有冲突。固然可以把`async`作为一个特例，但从TC39的历史看，几乎没有这样的特例。

马后炮来说，如果有 async do expression，那么其实不需要异步箭头函数，直接 `() => async do {}` 即可。但现在太迟了。

总得来说，还是命名的需求不够强。我的意思是，你总能通过单独把箭头函数赋值一个变量来解决（无论是自引用、递归还是调试所需的名称），或者包一层来命名（`{name: () => {}}.name`）尽管写起来有点不便，但这使得命名箭头函数成为了一种价值有限的语法糖（开销是固定的，而不像某些语法糖可能省略了平方级甚至指数级的开销），从而难以得到强力的支持。
